### PR TITLE
[BugFix]Fix select empty cloudnative table bug

### PR DESCRIPTION
**Pull request by abc982627271**
Fixes #issue

add a cn into cluster;
create a cloudnative table t;
execute "select * from t";
report error : "Backend node not found. Check if any backend node is down.";

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Review by kevincai — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java [hunk 1]
```diff
@@ -778,7 +778,7 @@ private void computeFragmentHosts() throws Exception {
             if (params.instanceExecParams.isEmpty()) {
                 Reference<Long> backendIdRef = new Reference<>();
                 TNetworkAddress execHostport;
-                if (usedComputeNode) {
+                if (usedComputeNode || RunMode.getCurrentRunMode() == RunMode.SHARED_DATA) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java:781]: what happen if the cluster only has backends?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|1|2|50.00%|[783]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 25 / 60 (41.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|16|49|32.65%|[556, 560, 561, 562, 563, 564, 565, 566, 796, 797, 798, 799, 800, 801, 802, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2192, 2193, 2194, 2195, 2196, 2197, 2198, 2199, 2200]|
|:large_blue_circle:|com/starrocks/qe/SimpleScheduler.java|8|10|80.00%|[91, 103]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 24 / 37 (64.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|15|26|57.69%|[527, 528, 529, 530, 531, 532, 539, 540, 550, 551, 552]|
|:large_blue_circle:|com/starrocks/qe/SimpleScheduler.java|8|10|80.00%|[92, 104]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java [hunk 2]
```diff
@@ -523,6 +523,38 @@ private List<List<TScanRangeParams>> splitScanRangeParamByRowCount(List<TScanRan
...
+        return SimpleScheduler.getBackendHost(this.idToBackend, nodeIdRef);
+    }
+
+    private TNetworkAddress getExecHostPortV2(long nodeId, List<TScanRangeLocation> locations,
+                                              Reference<Long> nodeIdRef) {
+        // TODO: need to refactor after be split into cn + dn
+        // get node from default warehouse
+        if (RunMode.getCurrentRunMode() == RunMode.SHARED_DATA) {
+            Map<Long, ComputeNode> idToNode = getNodeMapInDefaultWarehouse();
+            return SimpleScheduler.getHost(nodeId, locations, idToBackend,
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java:551]: `idToBackend` is a subset of `idToNode`.

suggestions:
recover `SimpleScheduler.getHost` with one map parameter and update this function below.
```
if (runmode == shared data) {
    return SimpleScheduler.getHost(nodeId, locations, idToNode, nodeIdRef);
} else {
    return SimpleScheduler.getHost(nodeId, locations, idToBackend, nodeIdRef);
}
```

**Review by kevincai — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SimpleScheduler.java [hunk 3]
```diff
@@ -76,38 +75,31 @@ public class SimpleScheduler {
...
         }
-        LOG.debug("getHost backendID={}, backendSize={}", backendId, backends.size());
+        LOG.debug("getHost nodeID={}, nodeSize={}", nodeId, computeNodes.size());
 
         // TODO: need to refactor after be split into cn + dn
-        ComputeNode backend = backends.get(backendId);
-        if (backend == null && RunMode.getCurrentRunMode() == RunMode.SHARED_DATA) {
-            backend = computeNodes.get(backendId);
-        }
+        ComputeNode node = computeNodes.get(nodeId);
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/qe/SimpleScheduler.java:89]: remove TODO

File: fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java [hunk 4]
```diff
@@ -195,8 +195,16 @@ public CoordinatorPreprocessor(TUniqueId queryId, ConnectContext context, List<P
         this.descriptorTable = null;
         this.fragments = fragments;
 
-        this.idToBackend = GlobalStateMgr.getCurrentSystemInfo().getIdToBackend();
         this.idToComputeNode = buildComputeNodeInfo();
+        if (RunMode.getCurrentRunMode() == RunMode.SHARED_DATA) {
+            this.idToBackend = buildComputeNodeInfo();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java:200]: copyOf(idToComputeNode)

File: fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java [hunk 5]
```diff
@@ -195,8 +195,16 @@ public CoordinatorPreprocessor(TUniqueId queryId, ConnectContext context, List<P
...
+        } else {
+            Map<Long, ComputeNode> nodeMap = new HashMap<>();
+            for (long nodeId : GlobalStateMgr.getCurrentSystemInfo().getIdToBackend().keySet()) {
+                nodeMap.put(nodeId, GlobalStateMgr.getCurrentSystemInfo().getBackendOrComputeNode(nodeId));
+            }
+            this.idToBackend = ImmutableMap.copyOf(nodeMap);
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java:206]:
```suggestion
            this.idToBackend = GlobalStateMgr.getCurrentSystemInfo().getIdToBackend();;
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 43 / 62 (69.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|5|9|55.56%|[397, 398, 402, 403]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|22|36|61.11%|[201, 420, 432, 439, 441, 452, 453, 454, 455, 456, 457, 501, 1381, 1383]|
|:large_blue_circle:|com/starrocks/qe/SimpleScheduler.java|11|12|91.67%|[102]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

**Comment by abc982627271** [comment]
@mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 35 / 54 (64.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/WarehouseManager.java|0|5|00.00%|[96, 97, 98, 99, 100]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|5|9|55.56%|[397, 398, 402, 403]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|14|23|60.87%|[200, 415, 423, 430, 432, 443, 487, 1367, 1369]|
|:large_blue_circle:|com/starrocks/qe/SimpleScheduler.java|11|12|91.67%|[101]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

**Review by imay — APPROVED** [review]

**Comment by imay** [comment]
@mergify backport branch-3.1

**Pull request merged**